doc/crypto/api/ops/key-agreement.rst

```diff
@@ -350,6 +356,276 @@ Combining key agreement and key derivation
+#.  Call `psa_key_agreement_iop_setup()` to specify the algorithm, and provide the private key and the peer public key.
+#.  Call `psa_key_agreement_iop_complete()` to finish the key agreement and output the shared secret, until this function does not return :code:`PSA_OPERATION_INCOMPLETE`.
+#.  If an error occurs at any stage, or to terminate the operation early, call `psa_key_agreement_iop_abort()`.
+
+
+.. typedef:: /* implementation-defined type */ psa_key_agreement_iop_t
+
+    .. summary::
+        The type of the state data structure for an interruptible key agreement operation.
+
+    Before calling any function on an interruptible key agreement operation object, the application must initialize it by any of the following means:
+
+    *   Set the object to all-bits-zero, for example:
+
+        .. code-block:: xref
+
+            psa_key_agreement_iop_t operation;
+            memset(&operation, 0, sizeof(operation));
+
+    *   Initialize the object to logical zero values by declaring the object as static or global without an explicit initializer, for example:
+
+        .. code-block:: xref
+
+            static psa_key_agreement_iop_t operation;
+
+    *   Initialize the object to the initializer `PSA_KEY_AGREEMENT_IOP_INIT`, for example:
+
+        .. code-block:: xref
+
+            psa_key_agreement_iop_t operation = PSA_KEY_AGREEMENT_IOP_INIT;
+
+    *   Assign the result of the function `psa_key_agreement_iop_init()` to the object, for example:
+
+        .. code-block:: xref
+
+            psa_key_agreement_iop_t operation;
+            operation = psa_key_agreement_iop_init();
+
+    This is an implementation-defined type.
+    Applications that make assumptions about the content of this object will result in implementation-specific behavior, and are non-portable.
+
+.. macro:: PSA_KEY_AGREEMENT_IOP_INIT
+    :definition: /* implementation-defined value */
+
+    .. summary::
+        This macro evaluates to an initializer for an interruptible key agreement operation object of type `psa_key_agreement_iop_t`.
+
+.. function:: psa_key_agreement_iop_init
+
+    .. summary::
+        Return an initial value for an interruptible key agreement operation object.
+
+    .. return:: psa_key_agreement_iop_t
+
+.. function:: psa_key_agreement_iop_get_num_ops
+
+    .. summary::
+        Get the number of *ops* that an interruptible key agreement operation has taken so far.
+
+    .. param:: psa_key_agreement_iop_t * operation
+        The interruptible key agreement operation to inspect.
+
+    .. return:: uint32_t
+        Number of *ops* that the operation has taken so far.
+
+    After the interruptible operation has completed, the returned value is the number of *ops* required for the entire operation.
+    The value is reset to zero by a call to either `psa_key_agreement_iop_setup()` or `psa_key_agreement_iop_abort()`.
```

Comment:
When we were discussing the specification, we implicitly assumed that setup takes negligible time. But this is not a safe assumption. For example, for key agreement, the public key needs to be validated. In Mbed TLS, we do this during setup (and it's the natural way to do it). Mbed TLS estimates the public key validation to take 3 ops. (The key agreement as a whole takes thousands of ops.)

So I think we should specify that setup resets the op count to 0 at the beginning of its operation, but then adds the number of ops that it takes. And recommend to implementations that this number should be lower than the value set with `psa_iop_set_max_ops`.

Reply:
This applies to all the functions, although key agreement might be the only one for which it's really natural to have setup take a few ops.

Reply:
That does complicate the way we can describe the reset of the _ops_ value, but is a valid point. Is it sufficient to expect that validation will occur during setup, or should the recommendation be that if validation can be very complex (substantially more _ops_ than a reasonable application-configured maximum) it should be deferred to the completion function?

Similar to the case for deferring key id allocation (`PSA_ERROR_ALREADY_EXISTS` returned by `psa_key_agreement_iop_complete()`), should we also permit `psa_key_agreement_iop_complete()` to return status codes related to the validation of the peer key? - which enables deferred validation.

Reply:
I think we should recommend that setup should only perform a small number of ops, preferably less than whatever was set by `psa_iop_set_max_ops()`.

And I hadn't thought of that, but indeed we should allow complete() to return status codes related to validations that may be considered slow.